File: content/problems/0309-best-time-to-buy-and-sell-stock-with-cooldown/solutions/state-machine-dp/solution.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int hold = -prices[0], sold = 0, rest = 0;
        for (int day = 1; day < (int)prices.size(); ++day) {
            int oldHold = hold, oldSold = sold, oldRest = rest;
            hold = max(oldHold, oldRest - prices[day]);
            sold = oldHold + prices[day];
            rest = max(oldRest, oldSold);
        }
        return max(sold, rest);
    }
};

File: content/problems/0309-best-time-to-buy-and-sell-stock-with-cooldown/solutions/state-machine-dp/Solution.java
class Solution {
    public int maxProfit(int[] prices) {
        int hold = -prices[0], sold = 0, rest = 0;
        for (int day = 1; day < prices.length; day++) {
            int oldHold = hold, oldSold = sold, oldRest = rest;
            hold = Math.max(oldHold, oldRest - prices[day]);
            sold = oldHold + prices[day];
            rest = Math.max(oldRest, oldSold);
        }
        return Math.max(sold, rest);
    }
}
